ark/phenotyping/cluster.py

```diff
@@ -46,6 +46,58 @@ def winner(sample, weights):
     return winning_coords
 
 
+def batch_winner(samples, weights):
+    """Predict multiple samples at once
+
+    Args:
+        samples (numpy.ndarray):
+            Contains the rows of the pixel matrix to predict
+        weights (numpy.ndarray):
+            A weight matrix of dimensions [num_x, num_y, num_chans]
+
+    Returns:
+        list:
+            The list with indices corresponding to the prediction of the respective sample
+    """
+
+    # only 1 sample provided: expand to 2-D
+    if len(samples.shape) == 1:
+        samples = np.expand_dims(samples, axis=0)
+
+    # collapse the weights for proper subtraction
+    # dimensions will be (m x n) x c
+    # (m x n): number of m and n SOM nodes
+    # c: number of channels
+    weights_collapse = np.reshape(weights, (weights.shape[0] * weights.shape[1], weights.shape[2]))
+
+    # subtract weights from each sample individually, need to reshape back to original
+    # for proper Euclidean distance calculation
+    samples_subtract = samples - weights_collapse[:, None]
+    samples_subtract = np.reshape(samples_subtract, (weights.shape[0], weights.shape[1],
+                                                     samples_subtract.shape[1],
+                                                     samples_subtract.shape[2]))
+
+    # take the Euclidean distance along the last axis, then reshuffle
+    # to a final matrix of r x (m x n)
+    # r: number of samples
+    # (m x n): number of m and n SOM nodes
+    activation_map = np.linalg.norm(samples_subtract, axis=-1)
```

Comment:
looked under the hood of `np.linalg.norm`.  for our purposes, we can bypass the 'sqrt' at the end here using:
```
activation_map = np.add.reduce(samples_subtract * samples_subtract, axis=-1)
```
since this won't change the winning coords.

Reply:
Good point

Reply:
shedding some light on the main slow down here.  The main slow down is actually not the norm/adding.  The main slowdown is actually the element wise product.  On a (10, 10, 100000, 30), the total time for norm (and add.reduce) is 780ms.  Almost **450ms** of that time is spent on the element-wise product!

Luckily, we can get ~5x speedup (170ms), using:
```
activation_map = np.einsum('ijkl,ijkl->ijk', samples_subtract, samples_subtract)
```

Furthermore, if we use np.float32's instead of 64's, we basically have a full 10x speedup (110ms).  Extrapolating to Candace's dataset, this would cluster in ~5mins as opposed to 12mins (but that should be tested).  If/when we add multithreading/processing, 4 threads tackle this on the order of 1 min.

Reply:
btw, resizing is free, time wise (like 1us levels of free)

Reply:
@ackagel yeah this should work, seems like `einsum` generally runs much faster than regular `numpy` multiplication due to better cache management and memory access in the underlying C implementation. I had originally complained about Sophie's C implementation of FlowSOM being very unintuitive, but now that I think about it, I think she did it that way to make the SOM training several orders more cache/memory efficient, which is in large part why Candace only needed 40 minutes to run it.

However, I'm not sure if we can extract this into our current MiniSOM-based training step. Fundamentally, the issue here is that the weights are updated per row, meaning we cannot calculate the winning coordinates for 100000 rows at a time without changing the way the algorithm trains. Hypothetically, we should get a similar speedup by structuring the array the way Sophie did it in Cython, which should also give us the C-optimization benefits as well. But at that point, it might be better to simply run Sophie's SOM directly.

@ngreenwald @ackagel what do you think?

Reply:
Yes, we can't do a weights update with 1 million rows at a time. What was the result of the R benchmarking? Is it fast on your laptop? If so, I think we should start with that, since it will likely require the least amount of work to just call the R function directly. Once we've got a working version, then I think it makes sense to think about whether we need to go back and optimize it ourself. 

My suspicion is that it would be one day's work at most to make the specific R functions we need callable from python. If that's not the case, then the above tradeoff might not be relevant.

Reply:
@ngreenwald hopefully I'll have the results by today, otherwise over the weekend. Been a pain setting up all the dependencies and configuring my R environment to run it.

Reply:
sounds good! no worries